ratelimit/middleware.py

```diff
@@ -9,6 +9,15 @@ def __init__(self, get_response):
         self.get_response = get_response
 
     def __call__(self, request):
+        """
+        Sets REMOTE_ADDR correctly to the client's ip, if REMOTE_ADDR is missing.
+        It happens when django is run with a domain socket server (eg. gunicorn).
+        ref: https://stackoverflow.com/a/34254843/1031191
+        """
+        if not request.META("REMOTE_ADDR", "") and "HTTP_X_FORWARDED_FOR" in request.META:
+            parts = request.META["HTTP_X_FORWARDED_FOR"].split(",", 1)
+            request.META["REMOTE_ADDR"] = parts[0]
```

Comment:
This is risky because it assumes too much about how the local proxies are configured.

It's possible that no proxy is in use, or other proxies may exist outside of your control. Even if you're aware of the local network infra, nefarious clients could prepend this, and since there is no clear standard for how to implement XFF, it's not clear what the reverse proxy server _should_ do here.

For example,

```http
GET /restricted-resource
Host: vulnerable.host
X-Forwarded-For: fake.ip.address, real.ip,address
```

From the [wiki article](https://en.wikipedia.org/wiki/X-Forwarded-For)

>  for this reason, effective use of XFF requires knowledge of which proxies are trustworthy, for instance by looking them up in a whitelist of servers whose maintainers can be trusted. 

In general, you want the right-most IP address that you don't recognize. 

```py
# settings.py
# proxies that exist in my infrastructure
KNOWN_PROXIES = [
    '10.1.2.3',
    '10.1.2.4',
    '10.1.2.5',
]

# middleware.py
class XFFMiddleware():
    def __call__(self, request):
        xff = request.META['HTTP_X_FORWARDED_FOR']
        ips = [ip.strip() for ip in xff.split(',')]
        for ip in reversed(ips):
            if ip not in settings.KNOWN_PROXIES:
                request.META['REMOTE_ADDR'] = ip
                break
        return self.get_response(request)
```

Reply:
Thanks @jsocol :O wow the internet can be so counter-intuitive sometimes.
I agree with you not to merge it, as you also said it in the docs. Also, your example could go to docs like a cookbook. I do site-wide and per user ratelimiting because ip-s are so easy to come by anyway.

Setting `proxy_set_header X-Real-IP $remote_addr;` in nginx seems to be the way to go, 
then using `key='header:real-ip'`. I think that is in the docs.

Reply:
If you’re ok running a pre-release version, check out what’s on the `main` branch: a new setting called `RATELIMIT_IP_META_KEY`. It would let you use settings (which are easier to override for local development) to tell ratelimit to look at the `HTTP_X_REAL_IP` entry in `request.META` instead of at `REMOTE_ADDR`. That way you could still use `key='ip'` or `user_or_ip`. 

Just be sure to match the configuration to how you configure nginx. And ensure that nginx is wiping out any existing values for that header.